distros/N/Net-AOLIM/TOC.7
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
'''   \*(M", \*(S", \*(N" and \*(T" are the equivalent of
'''   \*(L" and \*(R", except that they are used on ".xx" lines,
'''   such as .IP and .SH, which do another additional levels of
'''   double-quote interpretation
.ds M" """
.ds S" """
.ds N" """""
.ds T" """""
.ds L' '
.ds R' '
.ds M' '
.ds S' '
.ds N' '
.ds T' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds M" ``
.ds S" ''
.ds N" ``
.ds T" ''
.ds L' `
.ds R' '
.ds M' `
.ds S' '
.ds N' `
.ds T' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH TOC 7 "perl 5.004, patch 05" "6/Jul/2000" "User Contributed Perl Documentation"
.UC
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
TOC Documentation \- Documentation for the AOL TOC Instant Messenging Protocol
.SH "LICENSE"
Copyright (c) 1998-9 America Online, Inc. All Rights Reserved.
.PP
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.
.PP
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
.PP
You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place \- Suite 330, Boston, MA 02111-1307,
USA.
.SH "SYNOPSIS"
Version: TOC1.0
.PP
This document describes the protocol between TOC and TOC clients.  The
protocol is built on TCP.  Framing is done by SFLAP, described at the
bottom of this document.  Inside each SFLAP frame is a TOC command.
.PP
The TOC protocol is ASCII based, and special attention must be placed
argument separation.  The separator and the rules of separation are
different for messages inbound to TOC and outbound to the client.  The
rules of separation are described in sections below.
.PP
The TOC server is built mainly to service the TIC and TiK clients.
Since the TIC client is a Java applet, and downloadable, TOC will NOT
support multiple TOC protocol versions at the same time.  Therefore,
TiK users will be forced to upgrade if the protocol version changes.
TOC sends down the protocol version it expects the client to speak and
understand.  Note, the protocol version is a string.
.SH "IMPORTANT NOTES"
.Ip "\(bu \s-1TOC\s0 will drop the connection if a command exceeds the maximum   length, which is currently 1024 bytes.  So the client needs to spend   special attention to im, chat, and config message lengths.  There is   an 8k length maximum from \s-1TOC\s0 to the client." 4
.Ip "\(bu No commands should be sent to \s-1TOC\s0 (besides toc_signon) before   a \s-1SIGN_ON\s0 is received.  If you do send a command before \s-1SIGN_ON\s0 the   command will be ignored, and in some case the connection will be   dropped." 4
.Ip "\(bu Initial permit/deny items should be sent after receiving   \s-1SIGN_ON\s0 but before sending toc_init_done, otherwise the user will   flash on peoples buddylist who the user has denied.  You will   probably want to send the toc_add_buddies at this time also." 4
.Ip "\(bu After \s-1TOC\s0 sends the \s-1PAUSE\s0 message to a client, all messages   sent to \s-1TOC\s0 will be ignored, and in some cases the connection will   be dropped.  Another \s-1SIGN_ON\s0 message will be sent to the client when   it is online again.  The buddy list and permit/deny items must be   sent again, followed by the toc_init_done.  In most cases the   \s-1SIGN_ON\s0 message will be sent between 1-2 seconds after the \s-1PAUSE\s0   message.  Therefore a client could choose to ignore the \s-1PAUSE\s0   message and hope nothing bad happens." 4
.Ip "\(bu The string used for roasting passwords is \*(N'Tic/Toc\*(T' (no quotes), or \fB01010100011010010110001100101111010101000110111101100011\fR in binary." 4
.SH "DESCRIPTION"
.SH "Client \-> TOC"
The commands and the arguments are usually separated by whitespaces.
Arguments with whitespace characters should be enclosed in quotes.
Dollar signs, curly brackets, square brackets, parentheses, quotes,
and backslashes must all be backslashed whether in quotes or not.  It
is usually a good idea just to use quotes no matter what.  All user
names from clients to TOC should be normalized (spaces removed and
lowercased), and therefore are the one exception to the always use
quotes rule.
.PP
When sending commands to the server you will not get a response back
confirming that the command format was correct or not!  However in
some cases if the command format was incorrect the connection will be
dropped.
.Sh "toc_signon <authorizer host> <authorizer port> <User Name> <Password> <language> <version>"
.PP
.Vb 9
\&    The password needs to be roasted with the Roasting String if
\&    coming over a FLAP connection, CP connections don't use
\&    roasted passwords.  The language specified will be used
\&    when generating web pages, such as the get info pages.
\&    Currently the only supported language is 'english'.
\&    If the language sent isn't found, the default 'english'
\&    language will be used.  The version string will be used
\&    for the client identity, and must be less then 50
\&    characters.
.Ve
.Vb 7
\&    Passwords are roasted when sent to the host.  This is done so they 
\&    aren't sent in 'clear text' over the wire, although they are still 
\&    trivial to decode.  Roasting is performed by first xoring each byte 
\&    in the password with the equivalent modulo byte in the roasting 
\&    string.  The result is then converted to ascii hex, and prepended 
\&    with '0x'.  So for example the password 'password' roasts to 
\&    '0x2408105c23001130'
.Ve
.Sh "toc_init_done"
.PP
.Vb 6
\&    Tells TOC that we are ready to go online.  TOC clients should first 
\&    send TOC the buddy list and any permit/deny lists.  However toc_init_done
\&    must be called within 30 seconds after toc_signon, or the connection
\&    will be dropped.  Remember, it can't be called until after the SIGN_ON
\&    message is received.  Calling this before or multiple times after a
\&    SIGN_ON will cause the connection to be dropped.
.Ve
.Sh "toc_send_im <Destination User> <Message> [auto]"
.PP
.Vb 3
\&    Send a message to a remote user.  Remember to quote and encode the 
\&    message.  If the optional string 'auto' is the last argument, then the 
\&    auto response flag will be turned on for the im. 
.Ve
.Sh "toc_add_buddy <Buddy User 1> [<Buddy User2> [<Buddy User 3> [...]]]"
.PP
.Vb 2
\&    Add buddies to your buddy list.  This does not change your
\&    saved config.
.Ve
.Sh "toc_remove_buddy <Buddy User 1> [<Buddy User2> [<Buddy User 3> [...]]]"
.PP
.Vb 2
\&    Remove buddies from your buddy list.  This does not change your
\&    saved config.
.Ve
.Sh "toc_set_config <Config Info>"
.PP
.Vb 5
\&    Set the config information for this user.  The config information
\&    is line oriented with the first character being the item type,
\&    followed by a space, with the rest of the line being the item
\&    value.  Only letters, numbers, and spaces should be used.  Remember
\&    you will have to enclose the entire config in quotes.
.Ve
.Vb 11
\&    Item Types:
\&    g - Buddy Group (All Buddies until the next g or the end of config 
\&                     are in this group.)
\&    b - A Buddy 
\&    p - Person on permit list
\&    d - Person on deny list
\&    m - Permit/Deny Mode.  Possible values are
\&        1 - Permit All
\&        2 - Deny All
\&        3 - Permit Some
\&        4 - Deny Some
.Ve
.Sh "toc_evil <User> <norm|anon>"
.PP
.Vb 5
\&    Evil/Warn someone else.  The 2nd argument is either the string
\&    'norm' for a normal warning, or 'anon' for an anonymous 
\&    warning.  You can only evil people who have recently sent you
\&    ims.  The higher someones evil level, the slower they can
\&    send message.
.Ve
.Sh "toc_add_permit [ <User 1> [<User 2> [...]]]"
.PP
.Vb 6
\&    ADD the following people to your permit mode.  If
\&    you are in deny mode it will switch you to permit
\&    mode first.  With no arguments and in deny mode
\&    this will switch you to permit none. If already
\&    in permit mode, no arguments does nothing
\&    and your permit list remains the same.
.Ve
.Sh "toc_add_deny [ <User 1> [<User 2> [... ]]]"
.PP
.Vb 6
\&    ADD the following people to your deny mode. If
\&    you are in permit mode it will switch you to
\&    deny mode first.  With no arguments and in permit
\&    mode, this will switch you to deny none. If
\&    already in deny mode, no arguments does nothing
\&    and your deny list remains unchanged.
.Ve
.Sh "toc_chat_join <Exchange> <Chat Room Name>"
.PP
.Vb 13
\&    Join a chat room in the given exchange.  Exchange is
\&    an integer that represents a group of chat rooms.
\&    Different exchanges have different properties.  For
\&    example some exchanges might have room replication (ie
\&    a room never fills up, there are just multiple
\&    instances.) and some exchanges might have navigational
\&    information, and some exchanges might have ...  Currently
\&    exchange should always be 4, however this may
\&    change in the future.  You will either
\&    receive an ERROR if the room couldn't be joined
\&    or a CHAT_JOIN message.  The Chat Room Name
\&    is case insensitive and consecutive spaces
\&    are removed.
.Ve
.Sh "toc_chat_send <Chat Room \s-1ID\s0> <Message>"
.PP
.Vb 5
\&    Send a message in a chat room using the chat room
\&    id from CHAT_JOIN.  Since reflection is always on in
\&    TOC, you do not need to add the message to your chat UI,
\&    since you will get a CHAT_IN with the message.  
\&    Remember to quote and encode the message.
.Ve
.Sh "toc_chat_whisper <Chat Room \s-1ID\s0> <dst_user> <Message>"
.PP
.Vb 7
\&    Send a message in a chat room using the chat room
\&    id from CHAT_JOIN.  This message is directed at
\&    only one person.  (Currently you DO need to add this to
\&    your UI.)  Remember to quote and encode the message.  
\&    Chat whispering is different from IMs since it is linked
\&    to a chat room, and should usually be displayed in the chat
\&    room UI.
.Ve
.Sh "toc_chat_evil <Chat Room \s-1ID\s0> <User> <norm|anon>"
.PP
.Vb 3
\&    Evil/Warn someone else inside a chat room.  The 3rd argument is either 
\&    the string 'norm' for a normal warning, or 'anon' for an anonymous 
\&    warning.  Currently chat evil is not turned on in the chat complex.
.Ve
.Sh "toc_chat_invite <Chat Room \s-1ID\s0> <Invite Msg> <buddy1> [<buddy2> [<buddy3> [...]]]"
.PP
.Vb 2
\&    Once you are inside a chat room you can invite other people into
\&    that room.  Remember to quote and encode the invite message.
.Ve
.Sh "toc_chat_leave <Chat Room \s-1ID\s0>"
.PP
.Vb 1
\&    Leave the chat room.
.Ve
.Sh "toc_chat_accept <Chat Room \s-1ID\s0>"
.PP
.Vb 2
\&    Accept a CHAT_INVITE message from TOC.  The server will send a
\&    CHAT_JOIN in response.
.Ve
.Sh "toc_get_info <username>"
.PP
.Vb 2
\&    Gets a user's info a GOTO_URL or ERROR message will be sent back to the 
\&    client.
.Ve
.Sh "toc_set_info <info information>"
.PP
.Vb 2
\&    Set the LOCATE user information.  This is basic HTML.
\&    Remember to encode the info.
.Ve
.Sh "toc_set_away [<away message>]"
.PP
.Vb 5
\&    if the away message is present, then the unavailable
\&    status flag is set for the user.  If the away message
\&    is not present, then the unavailable status flag is
\&    unset.  The away message is basic HTML, remember to
\&    encode the information.
.Ve
.Sh "toc_get_dir <username>"
.PP
.Vb 2
\&    Gets a user's dir info a GOTO_URL or ERROR message will be sent back to the 
\&    client.
.Ve
.Sh "toc_set_dir <info information>"
.PP
.Vb 6
\&    Set the DIR user information.  This is a colon separated fields as
\&    in: 'first name':'middle name':'last name':'maiden
\&    name':'city':'state':'country':'email':'allow web searches' Should
\&    return a DIR_STATUS msg.  Having anything in the 'allow web
\&    searches' field allows people to use web-searches to find your
\&    directory info.  Otherwise, they'd have to use the client.
.Ve
.Sh "toc_dir_search <info information>"
.PP
.Vb 4
\&    Perform a search of the Oscar Directory, using colon separated
\&    fields as in: 'first name':'middle name':'last name':'maiden
\&    name':'city':'state':'country':'email' Returns either a GOTO_URL
\&    or ERROR msg.
.Ve
.Sh "toc_set_idle <idle secs>"
.PP
.Vb 5
\&    Set idle information. If <idle secs> is 0 then the user isn't idle at all.
\&    If <idle secs> is greater then 0 then the user has already been idle
\&    for <idle secs> number of seconds.  The server will automatically
\&    keep incrementing this number, so do not repeatedly call with new
\&    idle times.
.Ve
.SH "TOC \-> Client"
All user names from TOC to client are NOT normalized, and are sent as
they should be displayed.  String are NOT encoded, instead we use
colons as separators.  So that you can have colons inside of messages,
everything after the colon before :<Message> should be considered part
of the message (ie don't just \*(L'split\*(R' on colons, instead split with a
max number of results.)
.Sh "\s-1SIGN_ON\s0:<Client Version Supported>"
.PP
.Vb 3
\&   This is sent after a successful toc_signon command is sent to TOC.
\&   If the command was unsuccessful either the FLAP connection will
\&   be dropped or you will receive a ERROR message.
.Ve
.Sh "\s-1CONFIG\s0:<config>"
.PP
.Vb 3
\&   A user's config. Config can be empty in which case the host was not able to
\&   retrieve it, or a config didn't exist for the user.  See toc_set_config
\&   above for the format.
.Ve
.Sh "\s-1NICK\s0:<Nickname>"
.PP
.Vb 2
\&   Tells you your correct nickname (ie how it should be capitalized and
\&   spacing)
.Ve
.Sh "\s-1IM_IN\s0:<Source User>:<Auto Response T/F?>:<Message>"
.PP
.Vb 2
\&   Receive an IM from some one.  Everything after the third colon is
\&   the incoming message, including other colons.
.Ve
.Sh "\s-1UPDATE_BUDDY\s0:<Buddy User>:<Online? T/F>:<Evil Amount>:<Signon Time>:<IdleTime>:<\s-1UC\s0>"
.PP
.Vb 3
\&   This one command handles arrival/depart/updates.  Evil Amount is
\&   a percentage, Signon Time is UNIX epoc, idle time is in minutes, UC (User Class)
\&   is a two/three character string.
.Ve
.Vb 3
\&   uc[0]:
\&   ' '  - Ignore
\&   'A'  - On AOL
.Ve
.Vb 5
\&   uc[1]
\&   ' '  - Ignore
\&   'A'  - Oscar Admin
\&   'U'  - Oscar Unconfirmed
\&   'O'  - Oscar Normal
.Ve
.Vb 4
\&   uc[2] 
\&   '\e0' - Ignore
\&   ' '  - Ignore
\&   'U'  - The user has set their unavailable flag.
.Ve
.Sh "\s-1ERROR\s0:<Error Code>:Var args"
.PP
.Vb 5
\&   * General Errors *
\&   901   - $1 not currently available
\&   902   - Warning of $1 not currently available
\&   903   - A message has been dropped, you are exceeding
\&           the server speed limit
.Ve
.Vb 2
\&   * Chat Errors  *
\&   950   - Chat in $1 is unavailable.
.Ve
.Vb 4
\&   * IM & Info Errors *
\&   960   - You are sending message too fast to $1
\&   961   - You missed an im from $1 because it was too big.
\&   962   - You missed an im from $1 because it was sent too fast.
.Ve
.Vb 11
\&   * Dir Errors *
\&   970   - Failure
\&   971   - Too many matches
\&   972   - Need more qualifiers
\&   973   - Dir service temporarily unavailable
\&   974   - Email lookup restricted
\&   975   - Keyword Ignored
\&   976   - No Keywords
\&   977   - Language not supported
\&   978   - Country not supported
\&   979   - Failure unknown $1
.Ve
.Vb 8
\&   * Auth errors *
\&   980   - Incorrect nickname or password.
\&   981   - The service is temporarily unavailable.
\&   982   - Your warning level is currently too high to sign on.
\&   983   - You have been connecting and
\&           disconnecting too frequently.  Wait 10 minutes and try again.

distros/N/Net-AIM-TOC/TOC.pm
package Net::AIM::TOC;

$VERSION = '0.97';

use strict;

use Net::AIM::TOC::Config;

sub new {
	my $class = shift;

	my $self = {
		_conn => undef,
	};
	bless $self, $class;

	return( $self );
};

sub connect {
	my $self = shift;
	my $args = shift;

	my $conn = Net::AIM::TOC::Connection->new( $args );

	$self->{_conn} = $conn;

	return( 1 );
};

sub sign_on {
	my $self = shift;
	my $screenname = shift;
	my $password = shift;

	if( !defined($screenname) || !defined($password) ) {
		throw Net::AIM::TOC::Error( -text => 'Username/password not defined' );
	};

	my $ret = $self->{_conn}->send_signon( $screenname, $password );

	return( $ret );
};


sub send_im_to_aol {
	my $self = shift;
	my $user = shift;
	my $msg = shift;

	my $ret = $self->{_conn}->sendIMToAOL( $user, $msg );

	return( $ret );
};


sub send_to_aol {
	my $self = shift;
	my $msg = shift;

	my $ret = $self->{_conn}->sendToAOL( $msg );

	return( $ret );
};


sub recv_from_aol {
	my $self = shift;

	my( $msgObj ) = $self->{_conn}->recvFromAOL;

	return( $msgObj );
};


sub disconnect {
	my $self = shift;

	$self->{_conn}->disconnect;

	return( 1 );
};


=pod

=head1 NAME

Net::AIM::TOC - Perl implementation of the AIM TOC protocol
    
=head1 DESCRIPTION

The C<Net::AIM::TOC> module implements in AIM TOC protocol in such a way which make it simple for using when writing bots or AIM clients in Perl.

All of the code regarding the connection is abstracted in order to simplify the AIM connection down to merely sending and receiving instant messages and toc commands.

=head1 SYNOPSIS

  use Error qw( :try );
  use Net::AIM::TOC;

  try {
    my $aim = Net::AIM::TOC->new;
    $aim->connect;
    $aim->sign_on( $screenname, $password );

    my $msgObj = $aim->recv_from_aol;
    print $msgObj->getMsg, "\n";
    $aim->send_im_to_aol( $buddy, $msg );

    $aim->disconnect;
    exit( 0 );

  }
  catch Net::AIM::TOC::Error with {
    my $err = shift;
    print $err->stringify, "\n";

  };


=head1 CLASS INTERFACE

=head2 CONSTRUCTORS

A C<Net::AIM::TOC> object is created by calling the new constructor without arguments. A reference to the newly created object is returned, however, no connection to AIM has yet been made. One first is required to called C<connect> and C<sign_on> before attempting to send/receive instant messages.

=over 4

=item new ()

Returns C<Net::AIM::TOC> object but does not create a connection or sign on to the AIM service.

=back

=head2 OBJECT METHODS

=over 4

=item connect ( ARGS )

The connect method can be called without arguments to connect to the AIM service using the default AIM servers.

Alternatively, a hash containing any of the following keys can be passed in to connect to another service using the TOC protocol:

  -tocServer
  -tocPort
  -authServer
  -authPort

=item sign_on ( ARGS )

C<sign_on> is called to sign on to the AIM service. The arguments to be passed in are the screen name and password to be used to sign on to the service. 

=item send_im_to_aol ( ARGS )

Sends an instant message. The first argument should be the name of the receipient buddy and the second argument is the message which you are sending.

=item send_to_aol ( ARGS )

Sends whatever string is passed in on to the AIM service. Useful for sending toc commands.

=item recv_from_aol ()

Receives any data sent from the AIM service. This includes all TOC protocol messages (including instant messages), however, PAUSE And SIGN_ON messages are handled internally.

This method returns a C<Net::AIM::TOC::Messages> object. See the documentation for this object is to be used.

=item disconnect ()

Disconnects from the AIM service.

=back

=head1 KNOWN BUGS

None, but that does not mean there are not any.

=head1 SEE ALSO

C<Net::AIM::TOC::Messages>

=head1 AUTHOR

Alistair Francis, http://search.cpan.org/~friffin/

=cut


# Net::AIM::TOC::Connection package.
# Nothing to see here, please move along

package Net::AIM::TOC::Connection;

use strict;

use Net::AIM::TOC::Message;

use IO::Socket::INET;

sub new {
	my $class = shift;
	my $args = shift;

	my $self = {
		_sock	=> undef,
		_screenName	=> undef,
		_tocServer	=> $args->{tocServer} || Net::AIM::TOC::Config::TOC_SERVER,
		_tocPort	=> $args->{tocPort} || Net::AIM::TOC::Config::TOC_PORT,
		_authServer	=> $args->{authServer} || Net::AIM::TOC::Config::AUTH_SERVER,
		_authPort	=> $args->{authPort} || Net::AIM::TOC::Config::AUTH_PORT,
		_outseq	=> int(rand(100000)),
	};

	my $sock = IO::Socket::INET->new(
		PeerAddr	=> $self->{_tocServer},
		PeerPort	=> $self->{_tocPort},
		Type		=> SOCK_STREAM,
		Proto		=> 'tcp'
	);

	if( !defined($sock) ) {
		my $err_msg = 'Unable to connect to '. $self->{_tocServer} .' on port '. $self->{_tocPort};
		throw Net::AIM::TOC::Error( -text => $err_msg );
	};

	$self->{_sock} = $sock;
	bless $self, $class;

	return( $self );
};


sub send_signon {
	my $self = shift;
	my $screen_name = shift;
	my $password = shift;

	$self->{_screenName} = $screen_name;

	Net::AIM::TOC::Utils::printDebug( "send_signon: $screen_name" );

	my $data_out = "FLAPON\r\n\r\n";
	$self->{_sock}->send( $data_out );

	my( $msgObj ) = $self->recvFromAOL;
	Net::AIM::TOC::Utils::printDebug( $msgObj->getRawData );

	my $signon_data = pack "Nnna".length($screen_name), 1, 1, length($screen_name) , $screen_name;

	my $msg = pack "aCnn", '*', 1, $self->{_outseq}, length($signon_data);
	$msg .= $signon_data;

	my $ret = $self->{_sock}->send( $msg, 0 );

	if( !defined($ret) ) {
		throw Net::AIM::TOC::Error( -text => "syswrite: $!" );
	};

	my $login_string = $self->_getLoginString( $screen_name, $password );

	$ret = $self->sendToAOL( $login_string );

	# receive SIGNON data from AOL
	$msgObj = $self->recvFromAOL;
	Net::AIM::TOC::Utils::printDebug( $msgObj->getRawData );

	# Sending of sign on data is performed by 'recvFromAOL' to ensure
	# correct handling of PAUSE messages

	return( 1 );
};


sub _sendSignOnData {
	my $self = shift;

	# These lines are required in order to sign on
	my $ret = $self->sendToAOL( "toc_add_buddy $self->{_screenName}" );
	$ret = $self->sendToAOL( 'toc_set_config {m 1}' );

	# We're done with the signon process
	$ret = $self->sendToAOL( 'toc_init_done' );

	# remove the buddy we were required to add earlier
	$ret = $self->sendToAOL( "toc_remove_buddy $self->{_screenName}" );

	return;
};

sub _getLoginString {
	my $self = shift;
	my $screen_name = shift;
	my $password = shift;

	my $login_string = 'toc_signon '. $self->{_authServer} .' '. $self->{_authPort} .' '. $screen_name .' '. Net::AIM::TOC::Utils::encodePass( $password ) .' english '. Net::AIM::TOC::Utils::encode( Net::AIM::TOC::Config::AGENT );

	return( $login_string );
};


sub recvFromAOL {
	my $self = shift;

	my $buffer;

	if( !defined($self->{_sock}) ) {
		throw Net::AIM::TOC::Error( -text => 'We are not connected' );
	};

	my $ret = $self->{_sock}->recv( $buffer, 6 );
	if( !defined($ret) ) {
		throw Net::AIM::TOC::Error( -text => "sysread: $!" );
	};
	Net::AIM::TOC::Utils::printDebug( "RAW IN (header): '$buffer'" );

	my ($marker, $type, $in_seq, $len) = unpack "aCnn", $buffer;
	Net::AIM::TOC::Utils::printDebug( "IN (header): '$marker', '$type', '$in_seq', '$len'" );

	$ret = $self->{_sock}->recv( $buffer, $len );
	if( !defined($ret) ) {
		throw Net::AIM::TOC::Error( -text => "sysread: $!" );
	};
	Net::AIM::TOC::Utils::printDebug( "RAW IN (data): '$buffer'" );

	my $data = unpack( 'a*', $buffer );
	Net::AIM::TOC::Utils::printDebug( "IN (data): '$data'" );

	my $msgObj = Net::AIM::TOC::Message->new( $type, $data );

	if( $msgObj->getType eq 'SIGN_ON' ) {
		$self->_sendSignOnData;
	};

	return( $msgObj );
};


sub sendToAOL {
	my $self = shift;
	my $msg = shift;

	if( !defined($self->{_sock}) ) {
		throw Net::AIM::TOC::Error( -text => 'We are not connected' );
	};

	$msg .= "\0";

	Net::AIM::TOC::Utils::printDebug( "RAW OUT: $msg" );
	my $data = pack "aCnna*", '*', 2, ++$self->{_outseq}, length($msg), $msg;
	Net::AIM::TOC::Utils::printDebug( "OUT: $data" );

	my $ret = $self->{_sock}->send( $data, 0 );

	if( !defined($ret) ) {
		throw Net::AIM::TOC::Error( -text => "syswrite: $!" );
	};

	return( $ret );
};


sub sendIMToAOL {
	my $self = shift;
	my $user = shift;
	my $msg = shift;

	if( !defined($user) || !defined($msg) ) {
		Net::AIM::TOC::Utils::printDebug( "User or msg not defined\n" );
		return;
	};

	$user = Net::AIM::TOC::Utils::normalize( $user );
	$msg = Net::AIM::TOC::Utils::encode( $msg );

	$msg = 'toc_send_im '. $user .' '. $msg;

	my $ret = $self->sendToAOL( $msg );

	return( $ret );
};


sub disconnect {
	my $self = shift;

	$self->{_sock}->close;

	return;
};


# Net::AIM::TOC::Error* packages.
# Nothing to see here, please move along

package Net::AIM::TOC::Error;

use strict;

@Net::AIM::TOC::Error::ISA = qw( Error );


package Net::AIM::TOC::Error::Message;

use strict;

@Net::AIM::TOC::Error::Message::ISA = qw( Net::AIM::TOC::Error );



# Net::AIM::TOC::Utils package.
# Nothing to see here, please move along

package Net::AIM::TOC::Utils;

use strict;

sub printDebug {
	my $msg = shift;

	if( Net::AIM::TOC::Config::DEBUG ) {
		print STDERR $msg, "\n";
	};

	return;
};

sub encodePass {
	my $password = shift;

	my @table = unpack "c*" , 'Tic/Toc';
	my @pass = unpack "c*", $password;

	my $encpass = '0x';
	foreach my $c (0 .. $#pass) {
		$encpass.= sprintf "%02x", $pass[$c] ^ $table[ ( $c % 7) ];
	};

	return( $encpass );
};

sub encode {
	my $str = shift;

	$str =~ s/([\\\}\{\(\)\[\]\$\"])/\\$1/g;
	return( "\"$str\"" );
};

sub normalize {
	my $data = shift;
    
	$data =~ s/[^A-Za-z0-9]//g;
	$data =~ tr/A-Z/a-z/;

	return( $data );
};


sub removeHtmlTags {
	my $string = shift;
	my $replacement = shift || '';

	$string =~ s/<.*?>/$replacement/g;

	return( $string );
};


sub getCurrentTime {
	my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime(time);

	if( $sec < 10 ) { $sec = '0'.$sec };
	if( $min < 10 ) { $min = '0'.$min };

	return( "$hour:$min:$sec" );
};

1;

